File: sangsoo/0x08 utilization of stack/4949.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    while (true){
    string a;
    getline(cin, a);
    if (a == ".") break;
    bool flag = false;
    stack<char> stk;
    for (int i = 0; i < a.length(); i++){
        if ((a[i] == '(') || (a[i] == '[')){
            stk.push(a[i]);
        }
        else if (a[i] == ')'){
            if (!stk.empty() && stk.top() == '('){
                stk.pop();
            }
            else{
                cout << "no" << '\n';
                flag = true;
                break;
            }
        }
        else if (a[i] == ']'){
            if (!stk.empty() && stk.top() == '['){
                stk.pop();
            }
            else{
                cout << "no" << '\n';
                flag = true;
                break;
            }
        }
    }
    if (flag == true) continue;
    if(stk.empty()) cout << "yes" << '\n';
    else cout << "no" << '\n';
    }
}

File: sangsoo/0x08 utilization of stack/4949.py
import sys
while True:
    a = sys.stdin.readline().rstrip() #rstrip으로 해야함
    stk = []
    flag = False
    if a == '.': break
    for i in range(len(a)):
        if (a[i] == '(' or a[i] == '['):
            stk.append(a[i])
        elif a[i] == ')':
            if stk and stk[-1] == '(':
                stk = stk[:-1]
            else:
                print("no")
                flag = True
                break
        elif a[i] == ']':
            if stk and stk[-1] == '[':
                stk = stk[:-1]
            else:
                print("no")
                flag = True
                break
    if flag == True: continue
    else:
        if stk: print("no")
        else: print("yes")